File: src/frontend/cadastros/lib/app/data/model/pessoa_fisica_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:cadastros/app/infra/infra_imports.dart';
import 'package:cadastros/app/data/model/model_imports.dart';
import 'package:intl/intl.dart';
import 'package:cadastros/app/data/domain/domain_imports.dart';

class PessoaFisicaModel {
	int? id;
	int? idPessoa;
	int? idNivelFormacao;
	int? idEstadoCivil;
	String? cpf;
	String? rg;
	String? orgaoRg;
	DateTime? dataEmissaoRg;
	DateTime? dataNascimento;
	String? sexo;
	String? raca;
	String? nacionalidade;
	String? naturalidade;
	String? nomePai;
	String? nomeMae;
	EstadoCivilModel? estadoCivilModel;
	NivelFormacaoModel? nivelFormacaoModel;

	PessoaFisicaModel({
		this.id,
		this.idPessoa,
		this.idNivelFormacao,
		this.idEstadoCivil,
		this.cpf,
		this.rg,
		this.orgaoRg,
		this.dataEmissaoRg,
		this.dataNascimento,
		this.sexo,
		this.raca,
		this.nacionalidade,
		this.naturalidade,
		this.nomePai,
		this.nomeMae,
		this.estadoCivilModel,
		this.nivelFormacaoModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'cpf',
		'rg',
		'orgao_rg',
		'data_emissao_rg',
		'data_nascimento',
		'sexo',
		'raca',
		'nacionalidade',
		'naturalidade',
		'nome_pai',
		'nome_mae',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Cpf',
		'Rg',
		'Orgao Rg',
		'Data Emissao Rg',
		'Data Nascimento',
		'Sexo',
		'Raca',
		'Nacionalidade',
		'Naturalidade',
		'Nome Pai',
		'Nome Mae',
	];

	PessoaFisicaModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idPessoa = jsonData['idPessoa'];
		idNivelFormacao = jsonData['idNivelFormacao'];
		idEstadoCivil = jsonData['idEstadoCivil'];
		cpf = jsonData['cpf'];
		rg = jsonData['rg'];
		orgaoRg = jsonData['orgaoRg'];
		dataEmissaoRg = jsonData['dataEmissaoRg'] != null ? DateTime.tryParse(jsonData['dataEmissaoRg']) : null;
		dataNascimento = jsonData['dataNascimento'] != null ? DateTime.tryParse(jsonData['dataNascimento']) : null;
		sexo = PessoaFisicaDomain.getSexo(jsonData['sexo']);
		raca = PessoaFisicaDomain.getRaca(jsonData['raca']);
		nacionalidade = jsonData['nacionalidade'];
		naturalidade = jsonData['naturalidade'];
		nomePai = jsonData['nomePai'];
		nomeMae = jsonData['nomeMae'];
		estadoCivilModel = jsonData['estadoCivilModel'] == null ? EstadoCivilModel() : EstadoCivilModel.fromJson(jsonData['estadoCivilModel']);
		nivelFormacaoModel = jsonData['nivelFormacaoModel'] == null ? NivelFormacaoModel() : NivelFormacaoModel.fromJson(jsonData['nivelFormacaoModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idPessoa'] = idPessoa != 0 ? idPessoa : null;
		jsonData['idNivelFormacao'] = idNivelFormacao != 0 ? idNivelFormacao : null;
		jsonData['idEstadoCivil'] = idEstadoCivil != 0 ? idEstadoCivil : null;
		jsonData['cpf'] = Util.removeMask(cpf);
		jsonData['rg'] = rg;
		jsonData['orgaoRg'] = orgaoRg;
		jsonData['dataEmissaoRg'] = dataEmissaoRg != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataEmissaoRg!) : null;
		jsonData['dataNascimento'] = dataNascimento != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataNascimento!) : null;
		jsonData['sexo'] = PessoaFisicaDomain.setSexo(sexo);
		jsonData['raca'] = PessoaFisicaDomain.setRaca(raca);
		jsonData['nacionalidade'] = nacionalidade;
		jsonData['naturalidade'] = naturalidade;
		jsonData['nomePai'] = nomePai;
		jsonData['nomeMae'] = nomeMae;
		jsonData['estadoCivilModel'] = estadoCivilModel?.toJson;
		jsonData['nivelFormacaoModel'] = nivelFormacaoModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idPessoa = plutoRow.cells['idPessoa']?.value;
		idNivelFormacao = plutoRow.cells['idNivelFormacao']?.value;
		idEstadoCivil = plutoRow.cells['idEstadoCivil']?.value;
		cpf = plutoRow.cells['cpf']?.value;
		rg = plutoRow.cells['rg']?.value;
		orgaoRg = plutoRow.cells['orgaoRg']?.value;
		dataEmissaoRg = Util.stringToDate(plutoRow.cells['dataEmissaoRg']?.value);
		dataNascimento = Util.stringToDate(plutoRow.cells['dataNascimento']?.value);
		sexo = plutoRow.cells['sexo']?.value != '' ? plutoRow.cells['sexo']?.value : 'Masculino';
		raca = plutoRow.cells['raca']?.value != '' ? plutoRow.cells['raca']?.value : 'Branco';
		nacionalidade = plutoRow.cells['nacionalidade']?.value;
		naturalidade = plutoRow.cells['naturalidade']?.value;
		nomePai = plutoRow.cells['nomePai']?.value;
		nomeMae = plutoRow.cells['nomeMae']?.value;
		estadoCivilModel = EstadoCivilModel();
		estadoCivilModel?.nome = plutoRow.cells['estadoCivilModel']?.value;
		nivelFormacaoModel = NivelFormacaoModel();
		nivelFormacaoModel?.nome = plutoRow.cells['nivelFormacaoModel']?.value;
	}	

	PessoaFisicaModel clone() {
		return PessoaFisicaModel(
			id: id,
			idPessoa: idPessoa,
			idNivelFormacao: idNivelFormacao,
			idEstadoCivil: idEstadoCivil,
			cpf: cpf,
			rg: rg,
			orgaoRg: orgaoRg,
			dataEmissaoRg: dataEmissaoRg,
			dataNascimento: dataNascimento,
			sexo: sexo,
			raca: raca,
			nacionalidade: nacionalidade,
			naturalidade: naturalidade,
			nomePai: nomePai,
			nomeMae: nomeMae,
		);			
	}

	
}

File: backend/cadastros/src/model/pessoa_fisica_model.py
from src import db
from src.model.estado_civil_model import EstadoCivilModel
from src.model.nivel_formacao_model import NivelFormacaoModel


class PessoaFisicaModel(db.Model):
    __tablename__ = 'pessoa_fisica'

    id = db.Column(db.Integer, primary_key=True)
    cpf = db.Column(db.String(11))
    rg = db.Column(db.String(20))
    orgao_rg = db.Column(db.String(20))
    data_emissao_rg = db.Column(db.DateTime)
    data_nascimento = db.Column(db.DateTime)
    sexo = db.Column(db.String(1))
    raca = db.Column(db.String(1))
    nacionalidade = db.Column(db.String(100))
    naturalidade = db.Column(db.String(100))
    nome_pai = db.Column(db.String(200))
    nome_mae = db.Column(db.String(200))
    id_pessoa = db.Column(db.Integer, db.ForeignKey('pessoa.id'), unique=True)
    id_estado_civil = db.Column(db.Integer, db.ForeignKey('estado_civil.id'))
    id_nivel_formacao = db.Column(db.Integer, db.ForeignKey('nivel_formacao.id'))

    estado_civil_model = db.relationship('EstadoCivilModel', foreign_keys=[id_estado_civil])
    nivel_formacao_model = db.relationship('NivelFormacaoModel', foreign_keys=[id_nivel_formacao])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_pessoa = data.get('idPessoa')
        self.id_nivel_formacao = data.get('idNivelFormacao')
        self.id_estado_civil = data.get('idEstadoCivil')
        self.cpf = data.get('cpf')
        self.rg = data.get('rg')
        self.orgao_rg = data.get('orgaoRg')
        self.data_emissao_rg = data.get('dataEmissaoRg')
        self.data_nascimento = data.get('dataNascimento')
        self.sexo = data.get('sexo')
        self.raca = data.get('raca')
        self.nacionalidade = data.get('nacionalidade')
        self.naturalidade = data.get('naturalidade')
        self.nome_pai = data.get('nomePai')
        self.nome_mae = data.get('nomeMae')

    def serialize(self):
        return {
            'id': self.id,
            'idPessoa': self.id_pessoa,
            'idNivelFormacao': self.id_nivel_formacao,
            'idEstadoCivil': self.id_estado_civil,
            'cpf': self.cpf,
            'rg': self.rg,
            'orgaoRg': self.orgao_rg,
            'dataEmissaoRg': self.data_emissao_rg.isoformat(),
            'dataNascimento': self.data_nascimento.isoformat(),
            'sexo': self.sexo,
            'raca': self.raca,
            'nacionalidade': self.nacionalidade,
            'naturalidade': self.naturalidade,
            'nomePai': self.nome_pai,
            'nomeMae': self.nome_mae,
            'estadoCivilModel': self.estado_civil_model.serialize() if self.estado_civil_model else None,
            'nivelFormacaoModel': self.nivel_formacao_model.serialize() if self.nivel_formacao_model else None,
        }